### ISSUE TITLE
Adjustments: Improvements / Best Practices

### ISSUE THREAD

**Issue by solverat** [coreshop/CoreShop#945]
| Q                | A
| ---------------- | -----
| Bug report?      | no
| Feature request? | yes
| BC Break report? | no
| RFC?             | ~yes~ / closed

## Summary
- [ ] Documentation
- [ ]  Introduce a simple twig extension for adjustments
- [ ]  Introduce `$cart->hasAdjustments($type)`
- [x]  Add percentage/amount addition action
- [x]  Extend cart summary view, to allow positive / negative rules

***

## Documentation
There is currently no documentation, should be done with #641 

## Best Practice

> **Example**: As a ecommerce manager, i want to add some surcharges to the customers cart if the total amount does not reach a specific amount.

Currently, there is no nice/solid way to implement them, however, two approaches are possible:

### Solution I: Cart Rule
> This should be the preferred way - we're dealing with another rule here.
- Create a Cart Rule 
  - ✅ _possible_
- Add Amount Action 
  -  ✅ _possible_
- Add Amount Increase 
  - ⚠️  _missing_
- Display in Cart Summary 
  - ⚠️  _wrong output because every cart rule gets treated as a discount and will render a hard coded `-` symbol_

#### Possible To-Dos:
- Add `isNegative` flag to adjustment
- Add Amount/Percentage action to cart rule actions
- Calculate and display cart rule price modification based on `isNegative` flag.

***

### Solution II: Add Cart Processor

- Create a custom cart processor
  - ✅ _possible_
- Add your checks and create a custom Adjustment if necessary:
  -  ✅ _possible_

 ```php
$cart->addAdjustment(
    $this->adjustmentFactory->createWithData(
        self::SMALL_QUANTITIES_SURCHARGE,
        'Minimum Order Quantity',
        $surchargeWithoutTax,
        $surchargeWithTax
    )
);
```

- Check if adjustment has already been added to the cart
  - ⚠️  _missing -> need to filter all available adjustments_
- Show adjustment in cart summary
  - ⚠️ _missing -> need to generate a custom twig helper_

#### Possible To-Dos:
- Implement `$cart->getAdjustmentByType($type)`, `$cart->hasAdjustments($type)`
- Implement twig helper for adjustments

*** 

Personally, solution I is the way to go.

**Comment by dpfaffenbauer** [comment]
I would solve surcharges with a custom Cart Processor. So, on each process you remove all Adjustments of the type "SURCHARGES" and calculate if you need them, if you do, add them.

What is missing is how we display them in the cart summary correctly. I would introduce a new Twig Extension to get all adjustments of a certain type.

BTW: You can get all adjustments by a certain type: ```$cart->getAdjustments($type)``` returns an array with all adjustments of that type.

**Comment by dpfaffenbauer** [comment]
> Add isNegative flag to adjustment

Not needed, since adjustment can be negative or positive.

> Add Amount/Percentage action to cart rule actions

Yes, but that is not part of your issue/feature here.

> Calculate and display cart rule price modification based on isNegative flag.

Yes and no. A Twig Extension to get all adjustments of a certain type would do the job.

**Comment by solverat** [comment]
### Solution II - Discussed and approved for Implementation:
- Introduce Twig Extension
- Introduce `$cart->hasAdjustments($type)`

With that, Solution II has been cleared out.

### Solution I:
> I would solve surcharges with a custom Cart Processor. So, on each process you remove all Adjustments of the type "SURCHARGES" and calculate if you need them, if you do, add them.

So: CoreShop has cart rules and with that a lot of conditions. But I can't use them, because i need to add positive adjustments. 

Instead,  I need to:
- Implement all conditions by my own (processor)
- Calculate prices and taxes by my own (processor)
- Implement additional and nearly the same front-end markup as cart rules does

Right? This is just for clarification, if you have good arguments against it, we can skip the approach of using extended adjustments in cart rules.

>Yes, but that is not part of your issue/feature here.

It's what solution one is all about.

**Comment by dpfaffenbauer** [comment]
You can use Cart Rules, you just have to sort it yourself or create a PR for it. The only problem is how to calculate taxes for SURCHARGES?

I just wanted to state, you have both options: Cart Processor or Cart Rule (which in fact is a Cart Processor anyway). You just have to sort it somehow, and we can sort in the Core or you can sort in your Project, I am open for that.  There isn't much missing besides the Addition Amount/Addition Percent actions.

**Comment by solverat** [comment]
That's why this ticket is an RFC and that's why this ticket tries to clear things out. No one knew how to handle custom adjustments properly - until now.

Taxes:
- In a processor: Like you said: Do it however you want to do it, you're on your own -> use current context for example. But at least we could provide a documented example ("simple programmatic adjustment")
- In cart rule: Same as discount action

To sum things up:
- Introduce a simple twig extension for adjustments
- Introduce `$cart->hasAdjustments($type)`
- Add percentage/amount addition action
- Extend cart summary view, to allow positive / negative rules
- Of course, as usual, I'll provide some PRs for that

**Comment by dpfaffenbauer** [comment]
done

**Comment by dpfaffenbauer** [comment]
not completly done

**Comment by dpfaffenbauer** [comment]
This ticket is more than 2 years old now, I think it is fine as it is.
